JAVA-2959 Descriptive error when a Node has no available connections

akhaku opened the pull request:
For cases in which there are no connections available to
send requests to a Node in the query plan, collect the error
rather than silently skipping over the node. The error will
be thrown as part of an AllNodesFailedException if all nodes
fail.
This can happen when we've saturated the max in-flight
requests across all nodes or when the request is directed to
a particular node and it has no connections available (or
all its connections are saturated).
Note that in the latter case we used to throw a
NoNodeAvailableException but we now throw
AllNodesFailedException.

akhaku committed:
JAVA-2959 Descriptive error when a Node has no available connections

For cases in which there are no connections available to
send requests to a Node in the query plan, collect the error
rather than silently skipping over the node. The error will
be thrown as part of an AllNodesFailedException if all nodes
fail.
This can happen when we've saturated the max in-flight
requests across all nodes or when the request is directed to
a particular node and it has no connections available (or
all its connections are saturated).
Note that in the latter case we used to throw a
NoNodeAvailableException but we now throw
AllNodesFailedException.

akhaku commented:
Still needs tests for `CqlPrepareHandler` and `GraphRequestHandler`, but I wanted to get this up for feedback first. This pretty closely tracks Alex's guidance in the linked ticket but there's a technically breaking change i.e. a different exception type thrown in one case. The changes in ContinuousCqlRequestHandlerNodeTargetingTest illustrate the change in exception type.

akhaku commented on core/src/test/java/com/datastax/oss/driver/internal/core/cql/CqlRequestHandlerTest.java:
```diff
@@ -105,6 +110,28 @@ public void should_fail_if_no_node_available() {
+                      .handle();
+      assertThatStage(resultSetFuture)
+              .isFailed(error -> {
+                assertThat(error).isInstanceOf(AllNodesFailedException.class);
```
Clearly haven't worked with assertj before, I'd love feedback on how to make this block more readable!

adutra replied:
That's very well written, congrats :-)

Just as a suggestion, how about we include 2 nodes in this test? Something like:

```
@Test
public void should_fail_if_nodes_unavailable() {
  RequestHandlerTestHarness.Builder harnessBuilder = RequestHandlerTestHarness.builder();
  try (RequestHandlerTestHarness harness =
      harnessBuilder.withEmptyPool(node1).withEmptyPool(node2).build()) {
    CompletionStage<AsyncResultSet> resultSetFuture =
        new CqlRequestHandler(
                UNDEFINED_IDEMPOTENCE_STATEMENT,
                harness.getSession(),
                harness.getContext(),
                "test")
            .handle();
    assertThatStage(resultSetFuture)
        .isFailed(
            error -> {
              assertThat(error).isInstanceOf(AllNodesFailedException.class);
              Map<Node, List<Throwable>> allErrors =
                  ((AllNodesFailedException) error).getAllErrors();
              assertThat(allErrors).hasSize(2);
              assertThat(allErrors)
                  .hasEntrySatisfying(
                      node1,
                      nodeErrors ->
                          assertThat(nodeErrors)
                              .singleElement()
                              .isInstanceOf(NodeUnavailableException.class));
              assertThat(allErrors)
                  .hasEntrySatisfying(
                      node2,
                      nodeErrors ->
                          assertThat(nodeErrors)
                              .singleElement()
                              .isInstanceOf(NodeUnavailableException.class));
            });
  }
}
```

akhaku replied:
thanks, done

adutra commented on core/src/main/java/com/datastax/oss/driver/api/core/NodeUnavailableException.java:
```diff
@@ -0,0 +1,54 @@
+ * <p>This might happen under heavy load. The driver will automatically try the next node in the
+ * query plan. Therefore, the only way that the client can observe this exception is as part of a
+ * {@link AllNodesFailedException}.
+ * <p>The default load balancer will attempt to evenly distribute requests across nodes, so a common
```
You need to run `mvn fmt:format` on your branch, I see that a few classes are not compliant.

adutra replied:
That's a good explanation, but I wouldn't mention "load balancer" and "evenly distribute" here. I would rephrase this paragraph and the following one as follows:

```
 * <p>A common reason to encounter this error is when the configured number of connections per node
 * and requests per connection is not high enough to absorb the overall request rate. This can be
 * mitigated by tuning the following options:
 *
 * <ul>
 *   <li>{@code advanced.connection.pool.local.size};
 *   <li>{@code advanced.connection.pool.remote.size};
 *   <li>{@code advanced.connection.max-requests-per-connection}.
 * </ul>
 *
 * See {@code reference.conf} for more details.
 *
 * <p>Another possibility is when you are trying to direct a request {@linkplain
 * com.datastax.oss.driver.api.core.cql.Statement#setNode(Node) to a particular node}, but that node
 * has no connections available.
```

akhaku replied:
done and done

adutra commented on core/src/main/java/com/datastax/oss/driver/api/core/NodeUnavailableException.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Objects;
+
+/**
+ * Indicates that a Node encountered as part of processing a query plan has no connections
```
Just trying to improve the style a bit:

```suggestion
 * Indicates that a {@link Node} was selected in a query plan, but it had no connection available.
```

akhaku replied:
done, thanks

adutra reviewed (changes requested):
Impressive work, thank you!

Please format your code with `mvn fmt:format` and apply my suggestions if you agree with them.

Also, I would need an entry in the changelog, something like:

```
### 4.14.0 (in progress)

- [improvement] JAVA-2959: Don't throw NoNodeAvailableException when all connections busy
```

adutra commented:
> Still needs tests for `CqlPrepareHandler` and `GraphRequestHandler`, but I wanted to get this up for feedback first. This pretty closely tracks Alex's guidance in the linked ticket but there's a technically breaking change i.e. a different exception type thrown in one case. The changes in ContinuousCqlRequestHandlerNodeTargetingTest illustrate the change in exception type.

Tests for `CqlPrepareHandler` and `GraphRequestHandler` would be absolutely awesome, but don't spend too much time on them, basically the same test you wrote in `CqlRequestHandlerTest` would be more than enough.

And yes, I agree that this PR introduces a behavioral change. We might want to add a note in the upgrade guide as well and warn users that if they were expecting a `NoNodeAvailableException` in certain situations, they now will get an `AllNodesFailedException`.

akhaku commented:
Added a note to the upgrade guide as you suggested, ran `mvn fmt:format` and added a test to `CqlPrepareHandlerTest`. I started looking at adding a test to `GraphRequestHandlerTest` for completeness but it didn't look as easy as the other ones so I opted to take your advice and punt on it.

Thank you very much for the quick review! Sorry it was a busy week.

adutra commented on changelog/README.md:
```diff
@@ -2,6 +2,10 @@
 
 <!-- Note: contrary to 3.x, insert new entries *first* in their section -->
 
+### 4.14.0
```
```suggestion
### 4.14.0 (in progress)
```

akhaku replied:
done

adutra commented on upgrade_guide/README.md:
```diff
@@ -1,5 +1,15 @@
+
+#### AllNodesFailedException instead of NoNodeAvailableException in certain cases
+
+[JAVA-2959](https://datastax-oss.atlassian.net/browse/JAVA-2959) changed the behavior for when a
+request is directed to a particular node (via `Statement#setNode(Node)`) and all connections to that
+node are busy. Previously you would get back a `NoNodeAvailableException` but you will now get back
+an `AllNodesFailedException` where the `getAllErrors` map contains a `NodeUnavailableException`
+for that node.
```
`Statement#setNode(Node)` is not the only way by which you could observe the new behavior. I suggest a more generic rephrasing:

```suggestion
[JAVA-2959](https://datastax-oss.atlassian.net/browse/JAVA-2959) changed the behavior for when a
request cannot be executed because all nodes tried were busy. Previously you would get back a 
`NoNodeAvailableException` but you will now get back an `AllNodesFailedException` where the 
`getAllErrors` map contains a `NodeUnavailableException` for that node.
```

akhaku replied:
done

adutra reviewed (changes requested):
Thank you @akhaku we are almost there :-)

adutra reviewed (approved):
Awesome work, thank you!

akhaku commented:
Thank you! Please merge when ready - it looks like it's waiting for a branch check of some sort, perhaps that won't happen since this PR is from a fork.

adutra commented:
@akhaku would you mind signing our [CLA](https://cla.datastax.com/) if you haven't done so already? Thank you!

akhaku commented:
Gotcha, all done now
